### ISSUE TITLE
etcd rollback tool panics when the initial data comes from "etcdctl backup"

### ISSUE THREAD

**Issue by mml** [kubernetes/kubernetes#37131]
I have been performing some tests using a tarball taken from an apiserver filesystem and some tests using the output of `etcdctl-2.2.1 backup`.  In the latter case, the `rollback` tool will crash.

[data-crash.tar.gz](https://github.com/kubernetes/kubernetes/files/601123/data-crash.tar.gz) contains data that will crash `rollback`.  It is the data after having upgraded from 2.2.1 -> 2.3.7 -> 3.0.14.

[etcdctl-backup.tar.gz](https://github.com/kubernetes/kubernetes/files/601124/etcdctl-backup.tar.gz) contains both the original data (`etcd/data.orig`) and the output of `etcdctl-2.2.1 backup` (`etcd/data`) before any upgrades.

@wojtek-t @xiang90 @hongchaodeng

**Comment by mml** [comment]
And here's the stack trace.

```
+ /usr/local/bin/rollback --data-dir /var/etcd/data
panic: runtime error: slice bounds out of range

goroutine 1 [running]:
panic(0xc26260, 0xc820010030)
        /usr/local/go/src/runtime/panic.go:481 +0x3e6
k8s.io/kubernetes/vendor/github.com/coreos/etcd/wal.(*WAL).ReadAll(0xc82025c750, 0xc8202ae2c0, 0x12, 0x20, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/k8s.io/kubernetes/vendor/github.com/coreos/etcd/wal/wal.go:266 +0xc0d
main.rebuild(0x7fff3cd50f18, 0xe, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
        /go/src/k8s.io/kubernetes/cluster/images/etcd/rollback/rollback.go:251 +0x420
main.main()
        /go/src/k8s.io/kubernetes/cluster/images/etcd/rollback/rollback.go:103 +0x70c
```

**Comment by xiang90** [comment]
@wojtek-t @mml It is a dependency problem. We built rollback tool with etcd 3.0.x wal pkg (which is vendored by k8s), which generates wal files that cannot be opened by 2.3.7. Are we OK to ship a rollback binary which is built against 2.3.7 WAL pkg?

**Comment by mml** [comment]
@xiang90 I don't understand.  The `rollback` tool here is the thing that is crashing.  It is working using data that was generated by etcd 3.0.14 and it was also built against the 3.0.x tree.

#37035 is different.  That is etcd 2.3.7 crashing, and the explanation you give here makes sense for that bug.

**Comment by xiang90** [comment]
@mml 

As I said, the root cause is the same we use wrong WAL pkg. Either that WAL pkg fails or it generates bad output.

**Comment by mml** [comment]
@xiang90 have you confirmed this by re-building the rollback tool against 2.3.7?  If not, I will build the binary to verify it fixes the problem.

**Comment by xiang90** [comment]
@mml You can give it a try by pulling in 2.3.7 WAL pkg.

**Comment by wojtek-t** [comment]
Sorry - I'm lost in this bug. Let me clarify few thing and we will see what I'm missing.

The experiment Matt did are:
- he started 2.2.1 cluster
- then he migrated to 2.3.7
- then to 3.0.14
- then to 3.0.14 with v3 api
- then he tried to rollback and this crashed
@mml - please correct me if I'm wrong.

So the rollback tool was run on the 3.0.14 data. So @xiang90 - why are you saying that rollback tool tries to operate on 2.3.7 data?

**Comment by xiang90** [comment]
Rollback tools does two thing:

1. read out existing data
2. write out old data

The WAL pkg has to be able to read out existing data. So if you use a old WAL pkg, then panics can happen. If you use a new WAL pkg, then etcd can be confused. At somepoint in etcd3 release WAL can write/read both 2.3.7 and that version. Now it cannot. We need to ping the write out WAL to 2.3.7 WAL and read in WAL to the lasted etcd WAL.

**Comment by xiang90** [comment]
Basically one issue is about not able to read in WAL due to WAL version too low, one issue is about not able to read out the rollbacked data since WAL version is too high.

Once I use 2.3.7 WAL to write out data and 3.0.14 WAL to read in the data, both of the issues go away.

**Comment by wojtek-t** [comment]
OK - I think I roughly got the explanation. However I don't understand this:
```
 Now it cannot. We need to ping the write out WAL to 2.3.7 WAL and read in WAL to the lasted etcd WAL.
```

Also - can you please clarify exactly what the rollback procedure is in that case?
1. We have 3.0.14 in v3 api data.
2. Then we run "rollback tool" (how should we build it)?
3. Then we would like to start 2.3.7 from the rollbacked data.

**Comment by xiang90** [comment]
```
	"github.com/coreos/etcd/wal"
	"github.com/coreos/etcd/wal/walpb"
	wal237 "github.com/coreos/etcd237/wal"
```

Then use wal237 to do all writes, and use whatever we vendor to do read. I am not sure how we should do this in k8s repo. Probably we can change pull in wal237 as:

```
wal237 "k8s/rollback/etcd237/wal"
```

**Comment by wojtek-t** [comment]
@xiang90 - thanks.
@mml So it seems that we need to change this code:
https://github.com/kubernetes/kubernetes/blob/master/cluster/images/etcd/rollback/rollback.go

To what @xiang90 suggested (not yet sure how exactly to do it).

**Comment by xiang90** [comment]
@wojtek-t @mml 

Also note that the data presented here [data-crash.tar.gz](https://github.com/kubernetes/kubernetes/files/601123/data-crash.tar.gz)  is problematic itself.

@mml mentioned it is generated by etcd 3.0.14, but I cannot use 3.0.14 etcd to open it. So it becomes bad even before sending to the rollback tool somehow. But I cannot reproduce it by upgrading etcd with the same workflow described above.

I stated that both of the problems are resolved by changing WAL version, since:

1. I verified https://github.com/kubernetes/kubernetes/issues/37035 is fixed

2. I verified the workflow @mml mentioned in this pr wont work before the change, but works after the change.

**Comment by mml** [comment]
@wojtek-t thanks for pushing the discussion along.  I can hack together a build on my workstation to verify that I can fix the problem.  If I can get the tool to work, then we can discuss how to pull the frozen 2.3.7 code into our repo.

**Comment by mml** [comment]
@xiang90 and I spoke offline.  He suggested a modification like so:

```diff
diff --git a/cluster/images/etcd/rollback/rollback.go b/cluster/images/etcd/rollback/rollback.go
index ba75df4..3f09f77 100644
--- a/cluster/images/etcd/rollback/rollback.go
+++ b/cluster/images/etcd/rollback/rollback.go
@@ -39,6 +39,7 @@ import (
        "github.com/coreos/etcd/wal"
        "github.com/coreos/etcd/wal/walpb"
        "github.com/golang/glog"
+       wal237 "k8s.io/kubernetes/cluster/images/etcd/rollback/etcd/wal"
 )
 
 const rollbackVersion = "2.3.7"
@@ -115,7 +116,7 @@ func main() {
        }
        walDir := path.Join(*migrateDatadir, "member", "wal")
 
-       w, err := wal.Create(walDir, metadata)
+       w, err := wal237.Create(walDir, metadata)
        if err != nil {
                glog.Fatal(err)
        }
```

That builds and runs correctly, but as I mentioned in https://github.com/kubernetes/kubernetes/issues/37131#issuecomment-262036084, the crash is happening before this.  It's happening when `rollback` tries to read the files.

[etcd.log.txt](https://github.com/kubernetes/kubernetes/files/605040/etcd.log.txt) is the concatenated logs from all etcd runs.  Please note that by the time we run `rollback`, these should be 3.0.14 files.  Let us know if we're doing something wrong in the procedure here.

[migrate.log.txt](https://github.com/kubernetes/kubernetes/files/605041/migrate.log.txt) is verbose output from `migrate-if-needed.sh`.

**Comment by xiang90** [comment]
> @wojtek-t @mml Also note that the data presented here data-crash.tar.gz is problematic itself.
@mml mentioned it is generated by etcd 3.0.14, but I cannot use 3.0.14 etcd to open it. So ibecomes bad even before sending to the rollback tool somehow. But I cannot reproduce it by upgrading etcd with the same workflow described above.

@mml As I said, the data you provided is bad. I have no reason why it is bad.

I tried to use  [etcdctl-backup.tar.gz](https://github.com/kubernetes/kubernetes/files/601124/etcdctl-backup.tar.gz)  to upgrade etcd, everything works.

**Comment by xiang90** [comment]
>  It's happening when rollback tries to read the files.

Not only rollback tool cannot open this file. etcd 3.0.14 cannot even open this file. So I do not think this file is generated by etcd 3.0.14 or it is broken somehow. Try to do all the stuff again with your original data `etcdctl-backup.tar.gz`, then everything should be fine. I have tried, and it works.

**Comment by mml** [comment]
I tried again yesterday beginning with my original data and got exactly the same results.  Additionally, I started 3.0.14 with this data.  `kube-apiserver` contacted it and I ran `kubectl` to dump a list of all nodes and pods.  I am not sure why you aren't able to reproduce this.

@xiang90 can you please inspect the output from [etcd.log.txt](https://github.com/kubernetes/kubernetes/files/605040/etcd.log.txt) to see if we're doing something wrong?

Next step for me is to modify the upgrade script to produce a tarball of `/var/etcd/data/member` after each step.  I hope that with detailed snapshots of the filesystem, we can get on the same page about reproducibility.

**Comment by mml** [comment]
@xiang90 here are snapshots at every step.

- [snapshot-initial.tar.gz](https://github.com/kubernetes/kubernetes/files/607363/snapshot-initial.tar.gz)
- [snapshot-post-migrate-2.2.1.tar.gz](https://github.com/kubernetes/kubernetes/files/607362/snapshot-post-migrate-2.2.1.tar.gz)
- [snapshot-before-migrate-2.3.7.tar.gz](https://github.com/kubernetes/kubernetes/files/607364/snapshot-before-migrate-2.3.7.tar.gz)
- [snapshot-post-migrate-2.3.7.tar.gz](https://github.com/kubernetes/kubernetes/files/607365/snapshot-post-migrate-2.3.7.tar.gz)
- [snapshot-before-migrate-3.0.14.tar.gz](https://github.com/kubernetes/kubernetes/files/607361/snapshot-before-migrate-3.0.14.tar.gz)
- [snapshot-after-migrate-3.0.14.tar.gz](https://github.com/kubernetes/kubernetes/files/607359/snapshot-after-migrate-3.0.14.tar.gz)
- [snapshot-before-rollback-2.3.7.tar.gz](https://github.com/kubernetes/kubernetes/files/607360/snapshot-before-rollback-2.3.7.tar.gz)

**Comment by xiang90** [comment]
@mml Can you point me to a data dir that etcd 3.0.14 can successfully open, but a rollback tool that built with 3.0.14 WAL cannot open and panics?

To be honest, this sounds very strange to me. I cannot see a reason that the same WAL pkg behaviors differently in two different binaries. And it does not match my experiments. I guess we did something different.

**Comment by mml** [comment]
@xiang90 `after-migrate-3.0.14` is the state after migrating to 3.0.14.  From this I was able to start 3.0.14 and serve `kube-apiserver` requests.  After shutting 3.0.14 down, the state is as in `before-rollback-2.3.7`.  The files are not identical, but all that's happened in between is that I ran and shut down 3.0.14.

It is `before-rollback-2.3.7` that crashes.

**Comment by xiang90** [comment]
> After shutting 3.0.14 down

Can you still open the data-dir with etcd again?

> before-rollback-2.3.7

So you are able to open this with etcd 3.0.14 but not with a rollback tool built with 3.0.14's WAL pkg?

**Comment by mml** [comment]
> Can you still open the data-dir with etcd again?

Yes.  I just verified this.

> So you are able to open this with etcd 3.0.14 but not with a rollback tool built with 3.0.14's WAL pkg?

Yes.  The rollback tool is built with 3.0.14, except for the diff shown in https://github.com/kubernetes/kubernetes/issues/37131#issuecomment-262098529.

**Comment by mml** [comment]
If the exact steps help in debugging this, see #37309.  Particularly https://github.com/kubernetes/kubernetes/pull/37309/files#diff-df31784bf4f829f4fb8b2c74033d1f86.

**Comment by xiang90** [comment]
@mml 

With `before-rollback-2.3.7`, both of my etcd and rollback tool run well.

crash is untared `before-rollback-2.3.7`.

```
➜  rollback ./etcd-314 --data-dir=./crash
2016-11-22 13:17:15.249072 W | flags: unrecognized environment variable ETCDCTL_API=3
2016-11-22 13:17:15.249259 I | etcdmain: etcd Version: 3.0.14
2016-11-22 13:17:15.249267 I | etcdmain: Git SHA: 8a37349
2016-11-22 13:17:15.249273 I | etcdmain: Go Version: go1.6.3
2016-11-22 13:17:15.249278 I | etcdmain: Go OS/Arch: darwin/amd64
2016-11-22 13:17:15.249285 I | etcdmain: setting maximum number of CPUs to 4, total number of available CPUs is 4
2016-11-22 13:17:15.249430 W | etcdmain: found invalid file/dir data under data dir ./crash (Ignore this if you are upgrading etcd)
2016-11-22 13:17:15.251603 I | etcdmain: listening for peers on http://localhost:2380
2016-11-22 13:17:15.252078 I | etcdmain: listening for client requests on localhost:2379
2016-11-22 13:17:15.256408 I | etcdserver: name = default
2016-11-22 13:17:15.256441 I | etcdserver: data dir = ./crash
2016-11-22 13:17:15.256450 I | etcdserver: member dir = crash/member
2016-11-22 13:17:15.256456 I | etcdserver: heartbeat = 100ms
2016-11-22 13:17:15.256462 I | etcdserver: election = 1000ms
2016-11-22 13:17:15.256468 I | etcdserver: snapshot count = 10000
2016-11-22 13:17:15.256499 I | etcdserver: advertise client URLs = http://localhost:2379
2016-11-22 13:17:15.256511 I | etcdserver: initial advertise peer URLs = http://localhost:2380
2016-11-22 13:17:15.256534 I | etcdserver: initial cluster = default=http://localhost:2380
2016-11-22 13:17:15.636166 I | etcdserver: starting member 8e9e05c52164694d in cluster cdf818194e3a8c32
2016-11-22 13:17:15.637173 I | raft: 8e9e05c52164694d became follower at term 0
2016-11-22 13:17:15.637244 I | raft: newRaft 8e9e05c52164694d [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2016-11-22 13:17:15.637261 I | raft: 8e9e05c52164694d became follower at term 1
2016-11-22 13:17:15.644607 I | etcdserver: starting server... [version: 3.0.14, cluster version: to_be_decided]
2016-11-22 13:17:15.646258 E | etcdserver: cannot monitor file descriptor usage (cannot get FDUsage on darwin)
2016-11-22 13:17:15.648788 I | membership: added member 8e9e05c52164694d [http://localhost:2380] to cluster cdf818194e3a8c32
2016-11-22 13:17:15.846903 I | raft: 8e9e05c52164694d is starting a new election at term 1
2016-11-22 13:17:15.846966 I | raft: 8e9e05c52164694d became candidate at term 2
2016-11-22 13:17:15.846988 I | raft: 8e9e05c52164694d received vote from 8e9e05c52164694d at term 2
2016-11-22 13:17:15.847034 I | raft: 8e9e05c52164694d became leader at term 2
2016-11-22 13:17:15.847060 I | raft: raft.node: 8e9e05c52164694d elected leader 8e9e05c52164694d at term 2
2016-11-22 13:17:15.847919 I | etcdserver: setting up the initial cluster version to 3.0
2016-11-22 13:17:15.848465 I | etcdserver: published {Name:default ClientURLs:[http://localhost:2379]} to cluster cdf818194e3a8c32
2016-11-22 13:17:15.848489 I | etcdmain: ready to serve client requests
2016-11-22 13:17:15.852164 N | membership: set the initial cluster version to 3.0
2016-11-22 13:17:15.852258 I | api: enabled capabilities for version 3.0
2016-11-22 13:17:15.860641 N | etcdmain: serving insecure client requests on localhost:2379, this is strongly discouraged!
^C2016-11-22 13:17:18.718711 N | osutil: received interrupt signal, shutting down...

➜  rollback ./rollback --data-dir=crash
2016-11-22 13:17:27.913310 I | etcdserver/membership: added member 8e9e05c52164694d [http://localhost:2380] to cluster 0
2016-11-22 13:17:27.913632 N | etcdserver/membership: set the initial cluster version to 3.0
2016-11-22 13:17:27.913688 I | etcdserver/api: enabled capabilities for version 3.0
Finished successfully
➜  rollback
```

Also I manually checked `before-rollback-2.3.7`, it is very different from `data-crash.tar.gz` you provided in the original issue.

**Comment by mml** [comment]
Ah ha!  You are invoking rollback differently than we are.  See [migrate-if-needed.sh](https://github.com/kubernetes/kubernetes/blob/master/cluster/images/etcd/migrate-if-needed.sh#L235)

There is code in rollback.go that's supposed to detect the problem, but it's not working.  Tweaking things now.  Another update shortly.

**Comment by mml** [comment]
No.  That's not it.  But I see what the difference is.  We are removing the `*.snap` files first.  This was added in #36738 to fix #36555.

```console
% tar xfz log/snapshot-before-rollback-2.3.7.tar.gz; /tmp/tmp.JdHLvzwo9G/rollback --data-dir=./var/etcd/data 
2016-11-22 14:30:37.374728 I | membership: removed member cf7d617fc919f571 from cluster 0
2016-11-22 14:30:37.375082 I | membership: added member 588e22e5020001 [http://localhost:7001 http://localhost:2380] to cluster 0
2016-11-22 14:30:37.375278 N | membership: set the initial cluster version to 2.3
2016-11-22 14:30:37.375362 I | api: enabled capabilities for version 2.3
2016-11-22 14:30:37.375506 N | membership: updated the cluster version from 2.3 to 3.0
2016-11-22 14:30:37.375584 I | api: enabled capabilities for version 3.0
Finished successfully
```

But when I remove the `.snap` files.

```console
% tar xfz log/snapshot-before-rollback-2.3.7.tar.gz; rm -vf var/etcd/data/member/snap/*.snap; /tmp/tmp.JdHLvzwo9G/rollback --data-dir=./var/etcd/data
removed ‘var/etcd/data/member/snap/0000000000000002-000000000000ea66.snap’
panic: runtime error: slice bounds out of range

goroutine 1 [running]:
panic(0xc26260, 0xc820010140)
        /usr/local/go/src/runtime/panic.go:481 +0x3e6
k8s.io/kubernetes/vendor/github.com/coreos/etcd/wal.(*WAL).ReadAll(0xc82022d450, 0xc8202b9100, 0x12, 0x20, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/k8s.io/kubernetes/vendor/github.com/coreos/etcd/wal/wal.go:266 +0xc0d
main.rebuild(0x7ffcc4f5ee4f, 0xf, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
        /go/src/k8s.io/kubernetes/cluster/images/etcd/rollback/rollback.go:251 +0x420
main.main()
        /go/src/k8s.io/kubernetes/cluster/images/etcd/rollback/rollback.go:103 +0x70c
zsh: exit 2     /tmp/tmp.JdHLvzwo9G/rollback --data-dir=./var/etcd/data
```

**Comment by xiang90** [comment]
We can remove v2 data by using etcdctl, not remove the snapshot file.

**Comment by xiang90** [comment]
@mml Try not removing the snap file. The snapshot file is important.

**Comment by mml** [comment]
@xiang90 We're removing the snapshot file on the advice of @hongchaodeng (https://github.com/kubernetes/kubernetes/issues/36555#issuecomment-260224217).

Anyway, what's the correct etcdctl invocation to remove v2 data?

**Comment by wojtek-t** [comment]
The correct command is:
```
ETCDCTL_API=2 $ etcdctl rm <directory>
```
[I guess we can just do <directory>="registry" (or "/registry" I can't remember exactly).]

But for that we need the etcd to be up, IIUC.

**Comment by wojtek-t** [comment]
Once #37322 is merge, we should check if this one is fixed too.

**Comment by mml** [comment]
I'm not able to reproduce this now that we have the dual-WAL rollback tool including #37322.  Marking fixed, but the last little bit is #38253.
